nbgrader/apps/dbapp.py

```diff
@@ -116,48 +116,106 @@ def start(self):
+        raise NotImplementedError
+
+    @property
+    def description(self):
+        raise NotImplementedError
+
+    @property
+    def table_class(self):
+        raise NotImplementedError
+
+    @property
+    def primary_key(self):
+        """
+        This defaults to the primary key for the table_class. Override this if
+        the get_db_update_method does not expect its instance_id to be the
+        primary key.
+        """
+        possible_keys = [k.key for k in self.table_class.__mapper__.primary_key]
+        if len(possible_keys)>1:
+            raise TypeError("%s has more than one primary_key" % self.table_class.__name__)
+        else:
+            return possible_keys[0]
 
     def start(self):
-        super(DbStudentImportApp, self).start()
+        super(DbGenericImportApp, self).start()
 
         if len(self.extra_args) != 1:
             self.fail("Path to CSV file not provided.")
 
         path = self.extra_args[0]
         if not os.path.exists(path):
             self.fail("No such file: '%s'", path)
-        self.log.info("Importing students from: '%s'", path)
+        self.log.info("Importing from: '%s'", path)
 
-        allowed_keys = ["last_name", "first_name", "email", "id"]
 
         with Gradebook(self.coursedir.db_url) as gb:
             with open(path, 'r') as fh:
                 reader = csv.DictReader(fh)
+                reader.fieldnames = self._preprocess_keys(reader.fieldnames)
                 for row in reader:
-                    if "id" not in row:
-                        self.fail("Malformatted CSV file: must contain a column for 'id'")
+                    if self.primary_key not in row:
+                        self.fail("Malformatted CSV file: must contain a column for '%s'" % self.primary_key)
 
                     # make sure all the keys are actually allowed in the database,
                     # and that any empty strings are parsed as None
-                    student = {}
+                    instance = {}
                     for key, val in row.items():
-                        if key not in allowed_keys:
+                        if key not in self.allowed_keys:
                             continue
                         if val == '':
-                            student[key] = None
+                            instance[key] = None
                         else:
-                            student[key] = val
-                    student_id = student.pop("id")
+                            instance[key] = val
+                    instance_primary_key = instance.pop(self.primary_key)
+
+
+                    self.log.info("Creating/updating %s with %s '%s': %s",
+                                  self.table_class.__name__,
+                                  self.primary_key,
+                                  instance_primary_key,
+                                  instance)
+                    self.get_db_update_method(gb)(instance_primary_key, **instance)
```

Comment:
Rather than having this be a function you could use `getattr` and define the function name as a attribute of the class. For example:

```
class DbStudentImportApp(DbGenericImportApp):
    ...
    db_update_method = "update_or_create_student"
```

and then here you'd do `getattr(gb, self.db_update_method)(instance_primary_key, **instance)`

Reply:
good idea, done though I didn't do it in one line as that is more difficult to read and this is already decently complex.